typing/ctype.ml

```diff
@@ -54,7 +54,77 @@ open Btype
 
 (**** Errors ****)
 
-exception Unify of (type_expr * type_expr) list
+module Unification_trace = struct
+
+  type position = First | Second
+
+  type desc = { t: type_expr; expanded: type_expr option }
+  type 'a diff = { got: 'a; expected: 'a}
+
+  type 'a escape =
+    | Constructor of Path.t
+    | Univ of string option
+    | Self
+    | Module_type of Path.t
+    | Equation of 'a
+
+  type variant =
+    | No_intersection
+    | No_tags of position * (Asttypes.label * row_field) list
+    | Incompatible_types_for of string
```

Comment:
Mostly because I have an use case in mind for the type expressions in `Incompatible_fields`: adding an intermediary suberror message focused on the method. For instance,
> In method m, type 'a. 'a -> 'a is not compatible with 'a -> 'a
which would be quite useful for a new version of #1212 .

Contrarily, `Incompatible_types_for` (I should add `for_tag`) is an final element that is only used to explain a previous difference between the actual and expected types.